### PR TITLE
refactor(storage): return stream for state store iter

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Currently, we have our own trait `StateStoreIter`, which provides method `next` that returns a `StorageResult` holding the key value pair. Such interface is the same as a `Stream<Item = StorageResult<...>>`. For the following reasons, we may refactor our code to switch to return a stream instead of the current state store iter. 
- As we are moving the memtable down to the local state store, the `iter` method of local state store will return an iterator with lifetime. However, the lifetime of the current `StateStoreIter` must be `'static`.
- The current `next` call is the same as `try_next` of a `TryStream`. `Stream` is in the standard async library, and has many extension support in `StreamExt` and `TryStreamExt`, such as `map`, `collect`, `take`, for which we have manually support for `StateStoreIter`.
- The `next` method of a stream is cancellation safe, while the `next` of `StateStoreIter` isn't.

For the reasons above, we will return a `Stream` instead of `StateStoreIter`. To minimize the code change, the currently implemented `StateStoreIter` will be kept. We implement a method `into_stream` in `StateStoreIter`, which will turn the current `StateStoreIter` into a `Stream`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/storage/hummock_test/src/test_utils.rs [hunk 1]
```diff
@@ -18,7 +18,9 @@ use std::future::Future;
 use std::ops::Deref;
 use std::sync::Arc;
 
+use async_stream::try_stream;
```
Comment (BugenZhao) [inline src/storage/hummock_test/src/test_utils.rs:21]: `async_stream` is built with channels and performs worse than `futures_async_stream`. I guess we should prefer the latter.

Comment (wenym1) [reply src/storage/hummock_test/src/test_utils.rs:21]: I see. I will adopt it.

Comment (wenym1) [reply src/storage/hummock_test/src/test_utils.rs:21]: I have switched to use `futures_async_stream` in `risingwave_storage`. The `hummock_test` crates is still using `async_stream` for simplicity.

File: src/storage/hummock_test/Cargo.toml [hunk 2]
```diff
@@ -9,6 +9,7 @@ repository = { workspace = true }
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+async-stream = "0.3.3"
```
Comment (hzxa21) [inline src/storage/hummock_test/Cargo.toml:12]: Use `async-stream = "0.3"` to make it consistant with the specification in other packages.

File: src/storage/src/store_impl.rs [hunk 3]
```diff
@@ -338,7 +337,34 @@ pub mod verify {
...
+        let mut expected = expected.as_pin_mut();
+
+        loop {
+            let actual = actual.next().await;
+            if let Some(expected) = expected.as_mut() {
+                let expected = expected.next().await;
+                assert_eq!(actual.is_some(), expected.is_some());
+                if let Some(actual) = actual.as_ref() {
+                    assert_result_eq(actual, expected.as_ref().unwrap());
+                }
```
Comment (hzxa21) [inline src/storage/src/store_impl.rs:362]: Can be simplified to `assert_eq!(actual, expected)`?

**Review by hzxa21 — COMMENTED** [review]
Theoretically there should be no performance degradation with this change but I think we better verify it with a micro-bench. Maybe we can extend `bench_merge_iter`  to bench `StateStoreRead::iter`.

**Comment by wenym1** [comment]
Just add a micro-benchmark and run on the main branch and the current branch.
Main:
```
bench-hummock-iter      time:   [499.86 ms 500.87 ms 502.01 ms]
bench-hummock-iter      time:   [500.36 ms 501.29 ms 502.32 ms]
bench-hummock-iter      time:   [501.90 ms 503.24 ms 504.76 ms]
```

Current branch:
```
bench-hummock-iter      time:   [504.93 ms 505.53 ms 506.14 ms]
bench-hummock-iter      time:   [507.25 ms 507.88 ms 508.51 ms]
bench-hummock-iter      time:   [508.52 ms 509.58 ms 510.76 ms]
```

The performance of the current branch is slightly worse than the main branch, which is about 1%.

**Review by hzxa21 — APPROVED** [review]
LGTM

**Pull request merged**